File: solutions/elixir/sublist/1/lib/sublist.ex
defmodule Sublist do
  @doc """
  Returns whether the first list is a sublist or a superlist of the second list
  and if not whether it is equal or unequal to the second list.
  """
  def compare(a, b) do
    cond do
      a == b -> :equal
      contains(a, b) -> :superlist
      contains(b, a) -> :sublist
      true -> :unequal
    end
  end

  def contains([], _b), do: false
  def contains([_ | rest] = a, b), do: List.starts_with?(a, b) or contains(rest, b)
end

File: solutions/lua/sublist/1/sublist.lua
local function startsWith(l1, l2)
  if #l2 > #l1 then return false end
  for i = 1, #l2 do
    if l1[i] ~= l2[i] then return false end
  end
  return true
end

local function contains(l1, l2)
  if #l1 == 0 then return false end
  return startsWith(l1, l2) or contains(table.pack(table.unpack(l1, 2)), l2)
end

return function(l1, l2)
  if #l1 == #l2 and startsWith(l1, l2) then return "equal" end
  if contains(l1, l2) then return "superlist" end
  if contains(l2, l1) then return "sublist" end
  return "unequal"
end
